test/core/client_channel/lb_policy/pick_first_test.cc

```diff
@@ -39,6 +47,78 @@ class PickFirstTest : public LoadBalancingPolicyTest {
  protected:
   PickFirstTest() : lb_policy_(MakeLbPolicy("pick_first")) {}
 
+  static RefCountedPtr<LoadBalancingPolicy::Config> MakePickFirstConfig(
+      bool shuffle_address_list) {
+    return MakeConfig(Json::FromArray({Json::FromObject({{
+        "pick_first",
+        Json::FromObject(
+            {{"shuffleAddressList", Json::FromBool(shuffle_address_list)}}),
+    }})}));
+  }
+
+  absl::string_view GetConnectingSubchannel(
+      absl::Span<const absl::string_view> addresses,
+      SourceLocation location = SourceLocation()) {
+    for (auto address : addresses) {
+      auto* subchannel = FindSubchannel(
+          address, ChannelArgs().Set(GRPC_ARG_INHIBIT_HEALTH_CHECKING, true));
+      EXPECT_NE(subchannel, nullptr)
+          << location.file() << ":" << location.line();
+      if (subchannel == nullptr) {
+        return absl::string_view();
+      } else if (subchannel->ConnectionRequested()) {
+        return address;
+      }
+    }
+    return absl::string_view();
+  }
+
+  // Gets order the addresses are being picked. Return type is void so
+  // assertions can be used
+  void GetOrderAddressesArePicked(
+      absl::Span<const absl::string_view> addresses,
+      std::vector<absl::string_view>* out_address_order) {
+    absl::string_view address;
+    SubchannelState* subchannel = nullptr;
+    out_address_order->clear();
+    while (addresses.size() > out_address_order->size()) {
```

Comment:
The logic here is written in a way that is a little hard to follow.  For example, reporting TRANSIENT_FAILURE is done at the start of the loop for the subchannel from the previous loop iteration, so it's hard to understand the order in which the actions occur for each subchannel.  Also, we are winding up calling `FindSubchannel()` twice, once in `GetConnectingSubchannel()` and then again here, and we're doing this for every iteration of the loop.

Instead, I suggest constructing a list of subchannels at the start of the function, and removing elements from it as you iterate.  I think this approach results in more readable code:

```
// Construct a map of subchannel to address.
// We will remove entries as each subchannel starts to connect.
std::map<SubchannelState*, absl::string_view> subchannels;
for (auto address : addresses) {
  auto* subchannel = FindSubchannel(
      address, ChannelArgs().Set(GRPC_ARG_INHIBIT_HEALTH_CHECKING, true));
  ASSERT_NE(subchannel, nullptr);
  subchannels.emplace_back(subchannel, address);
}
// Now process each subchannel in the order in which pick_first tries it.
while (!subchannels.empty()) {
  // Find the subchannel that is being attempted.
  SubchannelState* subchannel = nullptr;
  for (const auto& p : subchannels) {
    if (p.first->ConnectionRequested()) {
      out_address_order->push_back(p.second);
      subchannel = p.first;
      break;
    }
  }
  ASSERT_NE(subchannel, "");
  // The subchannel reports CONNECTING.
  subchannel->SetConnectivityState(GRPC_CHANNEL_CONNECTING);
  // If this is the first subchannel being attempted, expect a CONNECTING update.
  if (subchannels.size() == kAddresses.size()) {
    ExpectConnectingUpdate();
  }
  if (subchannels.size() > 1) {
    // Not the last subchannel in the list.  Connection attempt should fail.
    subchannel->SetConnectivityState(
        GRPC_CHANNEL_TRANSIENT_FAILURE,
        absl::UnavailableError("failed to connect"));
    subchannel->SetConnectivityState(GRPC_CHANNEL_IDLE);
  } else {
    // Last subchannel in the list.  Connection attempt should succeed.
    subchannel->SetConnectivityState(GRPC_CHANNEL_READY);
    auto picker = WaitForConnected();
    ASSERT_NE(picker, nullptr);
    EXPECT_EQ(ExpectPickComplete(picker.get()), out_address_order->back());
    // Then it should become disconnected.
    subchannel->SetConnectivityState(GRPC_CHANNEL_IDLE);
    ExpectReresolutionRequest();
    ExpectStateAndQueuingPicker(GRPC_CHANNEL_IDLE);
  }
  // Remove the subchannel from the map.
  subchannels.erase(subchannel);
}
```